scipy/optimize/shgo_m/triangulation.py

```diff
@@ -0,0 +1,956 @@
+                # Return the wrapped function and wraps it to maintain the
+                # docstring and the name of the original function:
+                return functools.wraps(self._input_func)(return_func)
+
+            def __call__(self, *args, **kwargs):
+                """ Called for regular functions """
+                return self._cache_wrapper(None, *args, **kwargs)
+
+            # Set the cache_clear function in the __call__ operator:
+            __call__.cache_clear = cache_clear
+
+            def _cache_wrapper(self, caller, *args, **kwargs):
+                # Create a unique key including the types (in order to
+                # differentiate between 1 and '1'):
+                kwargs_key = "".join(map(
+                    lambda x: str(x) + str(type(kwargs[x])) + str(kwargs[x]),
+                    sorted(kwargs)))
+                key = "".join(
+                    map(lambda x: str(type(x)) + str(x), args)) + kwargs_key
+
+                # Check if caller exists, if not create one:
+                if caller not in self._caches_dict:
+                    self._caches_dict[caller] = [collections.OrderedDict(),
+                                                 time.time()]
+                else:
+                    # Validate in case the refresh time has passed:
+                    if self._timeout is not None:
+                        if (time.time() - self._caches_dict[caller][1]
+                                > self._timeout):
+                            self.cache_clear(caller)
+
+                # Check if the key exists, if so - return it:
+                cur_caller_cache_dict = self._caches_dict[caller][0]
+                if key in cur_caller_cache_dict:
+                    return cur_caller_cache_dict[key]
+
+                # Validate we didn't exceed the max_size:
+                if len(cur_caller_cache_dict) >= self._max_size:
+                    # Delete the first item in the dict:
+                    try:
+                        cur_caller_cache_dict.popitem(False)
+                    except KeyError:
+                        pass
+                # Call the function and store the data in the cache (call it
+                # with the caller in case it's an instance function
+                # - Ternary condition):
+                cur_caller_cache_dict[key] = self._input_func(caller, *args,
+                                                              **kwargs) if caller is not None else self._input_func(
+                    *args, **kwargs)
+                return cur_caller_cache_dict[key]
+
+        # Return the decorator wrapping the class (also wraps the instance to
+        # maintain the docstring and the name of the original function):
+        return (lambda input_func: functools.wraps(input_func)(
+            LruCacheClass(input_func, maxsize, timeout)))
+
+logging.basicConfig(stream=sys.stdout, level=logging.DEBUG)
```

Comment:
Libraries should not configure logging. I lost a chunk of time tracking down why my carefully-configured logging went haywire when I started using `shgo`. :-)

There is one exception: a root package's `__init__.py` can (ought to) add a `NullHandler` to that package's logger. This suppresses the "No handlers could be found for logger ..." message.

C.f. https://github.com/enthought/traits/blob/master/traits/__init__.py#L5-L22

scipy doesn't use `logging` anywhere else. You could place this in `scipy/optimize/__init__.py`

Reply:
Sorry about that! Fixed on the dev branch. Just waiting for the Travis test results.